### DISCUSSION TITLE
@testing-library/react-hooks not working with Zustand to update state

### DISCUSSION THREAD

**Discussion by Sup3r-Us3r** [pmndrs/zustand#1133] [Q&A]
I'm developing a lib for React/React Native where I export a hook to be used in the application, this hook is working correctly in a React/React Native application, I tested the lib locally.

Initially I used Context API, but it's a complex hook and I started having problems using the Context API, so I rewrote that hook using Zustand and it worked perfectly for what I needed, but the unit tests I had before working 100% with Context API, do not work in Zustand.

I'm using `@testing-library/react-hooks` to do the tests, but if I update some state and expect the result of that state to be the value I updated, it doesn't work.

`package.json`

```json
"jest": {
  "bail": true,
  "clearMocks": true,
  "collectCoverage": true,
  "collectCoverageFrom": [
    "<rootDir>/src/hooks/**/*.tsx",
    "<rootDir>/src/utils/*.ts"
  ],
  "coverageDirectory": "coverage",
  "coverageProvider": "v8",
  "coverageReporters": [
    "text",
    "lcov"
  ],
  "transform": {
    ".(ts|tsx)": "ts-jest"
  },
  "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
  "moduleFileExtensions": [
    "ts",
    "tsx",
    "js"
  ],
  "modulePathIgnorePatterns": [
    "<rootDir>/dist/"
  ]
},
"peerDependencies": {
  "react": ">= 17.0.2",
  "react-dom": ">= 17.0.2"
},
"devDependencies": {
  "@testing-library/react-hooks": "^8.0.0",
  "@types/jest": "^27.5.1",
  "@types/react": "^18.0.9",
  "@types/react-dom": "^18.0.4",
  "jest": "^28.0.3",
  "jest-environment-jsdom": "^28.1.0",
  "react": "^17.0.2",
  "react-dom": "^17.0.2",
  "react-test-renderer": "^18.1.0",
  "ts-jest": "^28.0.2",
  "typescript": "^4.6.4"
},
```

`useExample`

```ts
export const useExample = create(set => ({
  value: 0,
  add: (newValue: number) => set({ value: newValue }),
}))
```

`useExemple.spec.ts`

In this test where I was supposed to get 10, I get 0.

```ts
import { renderHook, act } from '@testing-library/react-hooks/server';

test('example', () => {
  const { result } = renderHook(() => useExemple());

  act(() => result.current.add(10));

  expect(result.current.value).toBe(10);
});
```

```ts
● useExemple tests › example

    expect(received).toBe(expected) // Object.is equality

    Expected: 10
    Received: 0

      15 |     act(() => result.current.add(10));
      16 |
    > 17 |     expect(result.current.value).toBe(10);
         |                                  ^
      18 |   });
      19 |
```

<br />

**obs**: I'm using `@testing-library/react-hooks/server` because for some reason if I use `@testing-library/react-hooks` it gives me the following error:

```ts
● useExample tests › example

    TypeError: (0 , _reactTestRenderer.act) is not a function

      11 | describe('useExample tests', () => {
      12 |   it('example', () => {
    > 13 |     const { result } = renderHook(() => useExample());
         |                                  ^
      14 |
      15 |     act(() => result.current.add(10));
      16 |

      at render (node_modules/@testing-library/react-hooks/lib/native/pure.js:73:34)
      at renderHook (node_modules/@testing-library/react-hooks/lib/core/index.js:114:5)
      at Object.<anonymous> (src/hooks/useExample/index.spec.tsx:13:34)
```

This error did not happen before with tests working in the Context API.

I really enjoyed working with Zustand, I managed to rewrite the hook very quickly, but the tests are important that they work, if anyone can help I would appreciate it.

**Comment by Yueran-Yu** [comment]
I have the same issue...

**Comment by Sup3r-Us3r** [comment]
@Yueran-Yu I managed to solve it, I don't know if it's the best solution, here's the step by step I did:

I installed the lib:

```
@testing-library/react
```

I removed the lib:

```
@testing-library/react-hooks
```

I needed to install the latest version of the `react` package and `react-dom` as development dependencies as versions like 17.0.2 didn't work.

```json
"react": "^18.1.0",
"react-dom": "^18.1.0",
```

In my jest config I added this:

```json
"globals": {
  "__DEV__": true
},
"testEnvironment": "jsdom",
```

In my `useExample.spec.ts` i added this:

```ts
import { renderHook, act } from '@testing-library/react';

describe('useExample tests', () => {
  beforeEach(() => {
    const { result } = renderHook(() => useExample());

    // Reset the state
    act(() => {
      result.current.value = 0;
    });
  });

  it('should be able to return an added number', () => {
    const { result } = renderHook(() => useExemple());
  
    act(() => result.current.add(10));
  
    expect(result.current.value).toBe(10);
  });
});
```

This worked for me, and it's what I'm using so far, obviously this is just an example, my current hook is quite complex so it has a lot more states and several functions, so I got the tests to work.

I even tested the solution to reset the hook state, according to this issue #271 but without success.

**Comment by Yueran-Yu** [comment]
> 

Thank you so much! I will try it out~

**Comment by Gustavohsdp** [comment]
I had the same problem :/

**Comment by charkour** [comment]
Thanks, @Sup3r-Us3r. 

Does anyone have a workaround for those currently on React 17 and can't update to React 18?

**Comment by charkour** [comment]
For React 17, I was able to solve my issue by wrapping the `expect` in a callback using the `it` function from Jest. @Sup3r-Us3r, would you be able to confirm this works with the old version before you upgraded? Thanks.

```tsx
import { renderHook, act } from '@testing-library/react-hooks';

test('example', () => {
  const { result } = renderHook(() => useExemple());

  act(() => result.current.add(10));

  it('is the right value', () => {
    // Note: you need to access `value` with `result.current`. If you destructure `value` from the object, it will become stale
    expect(result.current.value).toBe(10);
  });
});
```

**Reply by itsalb3rt** [reply]
You save my life!

**Comment by charkour** [comment]
Just kidding, that doesn't really work either.

**Comment by charkour** [comment]
Opted to use `store.getState()` in order to get the most recent value. The React 17 version of testing-library/react-hooks didn't seem to update properly.

**Comment by Sup3r-Us3r** [comment]
> Para o React 17, consegui resolver meu problema envolvendo o `expect`retorno de chamada usando a `it`função do Jest.@Sup3r-Us3r, você poderia confirmar que isso funciona com a versão antiga antes de atualizar? Obrigado.
> 
> ```tsx
> import { renderHook, act } from '@testing-library/react-hooks';
> 
> test('example', () => {
>   const { result } = renderHook(() => useExemple());
> 
>   act(() => result.current.add(10));
> 
>   it('is the right value', () => {
>     // Note: you need to access `value` with `result.current`. If you destructure `value` from the object, it will become stale
>     expect(result.current.value).toBe(10);
>   });
> });
> ```

Hi @charkour, I was developing a lib, and I needed to pass the React version inside `peerDependencies` I had no problems using the latest version of React, I tried to run the tests using version 17.0.2 and it didn't work unfortunately.

Your workaround in using an `it` inside the `test`, I've never tried it, I don't know if it's the best way around the problem.

Don't forget to clear the state values, as I did in my answer, otherwise your tests may fail.

**Comment by charkour** [comment]
> Your workaround in using an it inside the test, I've never tried it, I don't know if it's the best way around the problem.

I tested this some more, and this is not the best way to solve the problem. Sorry for my confusing suggestion. 

Thanks for reminding me to clear the state values! That was helpful.

I ended up using the vanilla version of the zustand store to test rather than the hook version.

**Comment by foxymiles** [comment]
I had the same problem. In my case I was able to fix it just by setting the jest environment to jsdom via a comment like this at the top of the test file:

```
/** @jest-environment jsdom */
```
